cgroups.c

```diff
@@ -0,0 +1,87 @@
+/* Copyright © 2020 Arista Networks, Inc. All rights reserved.
+ *
+ * Use of this source code is governed by the MIT license that can be found
+ * in the LICENSE file.
+ */
+
+#include <err.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <stdint.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "path.h"
+#include "cgroups.h"
+
+const struct cmap cgroup_map[] = {
+	{ CGROUPS_CLIMIT_CPU_MAX,          "cpu.max"          },
+	{ CGROUPS_CLIMIT_CPU_WEIGHT,       "cpu.weight"       },
+	{ CGROUPS_CLIMIT_MEMORY_MIN,       "memory.min"       },
+	{ CGROUPS_CLIMIT_MEMORY_LOW,       "memory.low"       },
+	{ CGROUPS_CLIMIT_MEMORY_HIGH,      "memory.high"      },
+	{ CGROUPS_CLIMIT_MEMORY_MAX,       "memory.max"       },
+	{ CGROUPS_CLIMIT_MEMORY_SWAP_HIGH, "memory.swap.high" },
+	{ CGROUPS_CLIMIT_MEMORY_SWAP_MAX,  "memory.swap.max"  },
+	{ CGROUPS_CLIMIT_MEMORY_RECLAIM,   "memory.reclaim"   },
+	{ CGROUPS_CLIMIT_IO_WEIGHT,        "io.weight"        },
+	{ CGROUPS_CLIMIT_IO_MAX,           "io.max"           },
+	{ CGROUPS_CLIMIT_IO_LATENCY,       "io.latency"       },
+	{ CGROUPS_CLIMIT_PIDS_MAX,         "pids.max"         },
+};
+
+/* Apply a limit, which contains a controller filename and quota value where cgroup_path
+ * specifies the name of the cgroup and cgroupfd is the open fd of the cgroup directory.
+ */
+void apply_climits(const struct climit * limits, const char * cgroup_path, const int cgroupfd) {
+	for (size_t i = 0; i < CGROUPS_NLIMIT; ++i) {
+		if (limits[i].present) {
+			int fd = openat(cgroupfd, limits[i].fname, O_WRONLY);
+			if (fd == -1) {
+				err(1, "unable to open %s at %s", limits[i].fname, cgroup_path);
+			}
+
+			if (write(fd, limits[i].clim, strlen(limits[i].clim)) == -1) {
+				err(1, "unable to write %s to %s/%s", limits[i].clim, cgroup_path, limits[i].fname);
+			}
+
+			close(fd);
+		}
+	}
+}
+
+/* Joing the cgroup provided by the cgroup_path name and the fd of the directory 
+ * that it resides in.
+ */
+void join_cgroup(const char * cgroup_path, const int cgroupfd) {
+	/* Open and join specified cgroups. */
+	int fd = openat(cgroupfd, "cgroup.procs", O_WRONLY);
+	if (fd == -1) {
+		err(1, "unable to open cgroup.procs at %s", cgroup_path);
+	}
+	const char *data = "0";
+	if (write(fd, data, 1) == -1) {
+		err(1, "joining cgroup %s", cgroup_path);
+	}
+	close(fd);
+}
+
+/* Called when bst terminates. This removes the created bst directory at the given
+ * cleanfd file descriptor.
+ */
+void cgroup_clean(int cleanfd, const char * cgroup_clean) {
+	join_cgroup(cgroup_clean, cleanfd);
```

Comment:
why is this needed?

Reply:
I have added a comment explaining this. bst cannot remove a cgroup if the bst pid is a member of that cgroup. This will give some "device or resource busy" error if one tries otherwise. Open to other ideas here though.